src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/AssessmentTransformer.kt

```diff
@@ -76,6 +85,23 @@ class AssessmentTransformer(
         personInfo,
       ) as TemporaryAccommodationApplication
 
+      val notes = jpa.referralHistoryNotes.map {
+        if (it.id == lastReferralRejectedHistoryNote?.id) {
+          assessmentReferralHistoryNoteTransformer.transformJpaToApi(it, jpa)
+        } else {
+          assessmentReferralHistoryNoteTransformer.transformJpaToApi(it)
+        }
+      }.toMutableList()
+
+      notes.addAll(
+        cas3Events.map {
+          val user = userService.findById(it.triggeredByUserId!!)
+            .orElseThrow { NotFoundProblem(it.triggeredByUserId!!, "UserEntity") }
+          val event = objectMapper.readValue<CAS3AssessmentUpdatedEvent>(it.data)
```

Comment:
IMO the only thing that should 'understand' the JSON in the domain events is the Domain Event Service to better encapsulate how we manage the serialization/deserialization.

Furthermore, this code is assuming that cas3Events only contains CAS3 Assessment Update Events but there's nothing at compile time validating this

I propose the following:

1. transformJpaToApi should accept List <CAS3AssessmentUpdatedEvent>
2. this list is provided by your domain event service (we do something similar in CAS1). You could maybe update domainEventService.getAssessmentUpdatedEvents to return a list of uk.gov.justice.digital.hmpps.approvedpremisesapi.model.DomainEvent instead of the entity? (you'll need to add 'triggeredByUserId' to that)

Reply:
Makes sense, and keeps the code generic to handle other domain event types here if required 👍